daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/validation/ValidationSpec.scala

```diff
@@ -0,0 +1,651 @@
+      "tweakExerciseExerciseResult(Some)" -> tweakExerciseExerciseResultSome,
+      "tweakExerciseKey(Some)" -> tweakExerciseKey(tweakOptKeyMaintainersSome),
+      "tweakExerciseByKey(New Version)" -> tweakExerciseByKey(versionSinceMinByKey),
+      "tweakExerciseVersion" -> tweakExerciseVersion,
+    )
+
+  private val insigExeTweaks =
+    Map(
+      "tweakExerciseOptLocation" -> tweakExerciseOptLocation,
+      "tweakExerciseExerciseResult(None)" -> tweakExerciseExerciseResultNone,
+      "tweakExerciseKey(None)" -> tweakExerciseKey(tweakOptKeyMaintainersNone),
+      "tweakExerciseByKey(Old Version)" -> tweakExerciseByKey(versionBeforeMinByKey),
+    )
+
+  //--[significant and insignificant tx tweaks]--
+
+  private def tweakTxNodes(tweakNode: Tweak[Node]) = Tweak[VTX] { vtx =>
+    // tweak any node in a transaction
+    vtx.transaction match {
+      case GenTransaction(nodeMapA, roots) =>
+        for {
+          nid <- nodeMapA.keys.toList
+          nodeB <- tweakNode.run(nodeMapA(nid))
+        } yield {
+          val nodeMapB = nodeMapA + (nid -> nodeB)
+          val txB = GenTransaction(nodeMapB, roots)
+          TransactionVersion.asVersionedTransaction(txB)
+        }
+    }
+  }
+
+  private def significantTweaks: Map[String, Tweak[VTX]] = {
+    (sigCreateTweaks ++ sigFetchTweaks ++ sigLookupTweaks ++ sigExeTweaks)
+      .map { case (name, tw) => (name, tweakTxNodes(tw)) }
+  }
+
+  private def insignificantTweaks: Map[String, Tweak[VTX]] = {
+    (insigCreateTweaks ++ insigFetchTweaks ++ insigLookupTweaks ++ insigExeTweaks)
+      .map { case (name, tw) => (name, tweakTxNodes(tw)) }
+  }
+
+  //--[per tweak tests]--
+
+  "Significant tweaks" - {
+    var count = 0
+    significantTweaks.foreach { case (name, tweak) =>
+      val pairs = runTweak(tweak)
+      val n = pairs.length
+      count += n
+      assert(n > 0) // ensure tweak actualy applies to something
+      s"[#$n] $name" in {
+        pairs.foreach { case (txA, txB) =>
+          val res = Validation.isReplayedBy(txA, txB)
+          res match {
+            case Left(_) => ()
+            case Right(()) =>
+              println(s"**expected $name to be SIGNIFICANT (but it wasn't)\n-txA=$txA\n-txB=$txB")
+              assert(false)
+          }
+        }
+      }
+    }
+    println(s"**SIGNIFICANT tweaks checked: #$count")
+  }
+
+  "Insignificant tweaks" - {
+    var count = 0
+    insignificantTweaks.foreach { case (name, tweak) =>
+      val pairs = runTweak(tweak)
+      val n = pairs.length
+      count += n
+      assert(n > 0) // ensure tweak actualy applies to something
+      s"[#$n] $name" in {
+        pairs.foreach { case (txA, txB) =>
+          val res = Validation.isReplayedBy(txA, txB)
+          res match {
+            case Right(()) => ()
+            case Left(_) =>
+              println(s"**expected $name to be INSIGNIFICANT (but it wasn't)\n-txA=$txA\n-txB=$txB")
+              assert(false)
+          }
```

Comment:
because it is quite helpful to be able to see the `txA` and `txB`

Reply:
:+1:

Reply:
`forEvery` solves this problem nicely, as mentioned in https://github.com/digital-asset/daml/pull/10167#discussion_r664890874.

But there's another aspect here that really illustrates why it's useful to track down the existing combinators rather than use scala-library to simulate them. There are two kinds of failures of these kinds of tests: break-everything, and break-something. As written, these tests stop on the first failure. So you don't really know which of those situations you're in.

With `forEvery`, you can immediately tell from one test run whether you've broken everything, just a few things, or a single edge case. It's a bit like the difference between having separate test declarations for each test, and just putting every assertion in the whole file into a single test case.

Redefining this functionality is _annoying_. And it is quite useful; #8879 was more or less an exercise of adding cases, tweaking or rewriting part of the algorithm, and quickly getting feedback on every shortcoming of whatever change I had just made--whether there was just one detail I had forgotten, or everything collapsed and I had hit a dead end.